terraform/cross-device/workload_identity.tf

```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module "gke-workload-identity" {
```

Comment:
You don't need to create additional service accounts. For each Kubernetes Namespace (and GKE node pool, because in this architecture there's 1:1 mapping between Kubernetes Namespaces and GKE node pools), there are two Cloud IAM service accounts: one for the nodes in that node pool, one for the workloads that run on those nodes.

Instead of creating a Cloud IAM service account for the workloads like you're doing here, reuse the Cloud IAM service account for the workloads as an input to your module, and grant the additional roles you need to that. Note that you'll need to know the namespace (and so the node pool) where you want to deploy this stuff because there's a [map](https://github.com/GoogleCloudPlatform/federated-learning/blob/5281bbfed272123cc41ee7025842856fca702a2b/terraform/main.tf#L25) that defines the names of all these service accounts.

Finally, you're leaving the creation of the Kubernetes service account to the `workload_identity` module. This is going to be an issue because you want to know the name of that service account to use it in your Kubernetes deployments. I suggest setting the `k8s_sa_name` variable to a known value that you can reuse later, instead of letting the module create the Kubernetes service account for you, or eventually reuse the Kubernetes service account that's already configured for workloads in that node pool. This is up to you.

Note: You can't add roles to the Cloud IAM service account we create for the GKE nodes because we enforce an [authoritative mode](https://github.com/GoogleCloudPlatform/federated-learning/blob/5281bbfed272123cc41ee7025842856fca702a2b/terraform/iam.tf#L33) on that.

Reply:
I think it's working, WDYT ?

Reply:
Is this working in its current form? Where are you passing the value of `var.google_service_account`?